content/blog/20200130-jepsen.md

```diff
@@ -0,0 +1,81 @@
+
+### Documentation Issues
+
+The Jepsen analysis does not only cover software correctness but also verifies the documentation claims. The Jepsen author found documentation issues and misleading use case examples of etcd.  The previous etcd documentation on consistency and isolation is not clear and might cause confusion.
+
+It describes the consistency model based on [the Wikipedia definition](https://en.wikipedia.org/wiki/Consistency_model), which is also the classification some etcd engineers [learned in University](http://www.cs.cmu.edu/~srini/15-446/S09/lectures/10-consistency.pdf). The previous documentation also separates the isolation level from consistency level, as there is no general agreement on [how the two should be defined together](https://fauna.com/blog/demystifying-database-systems-part-4-isolation-levels-vs-consistency-levels).
+
+The Jepsen author suggests etcd documentation adopt the term “strict serializability” originally defined in the [Linearizability paper published by Herlihy and Wing](https://cs.brown.edu/~mph/HerlihyW90/p463-herlihy.pdf). After reading the documentation on other open source projects, as well as listening to feedback from our users, the etcd team agrees with the Jepsen author that the [suggested model and term](https://jepsen.io/consistency) are clearer and easier to understand.
+
+Next, the lock API use case example is misleading and might cause runtime issues.
+
+Like any other distributed lock systems, an etcd lock provides different guarantees than say a local thread level lock. Specifically, an etcd lock only safely guarantees mutual exclusion inside etcd's own keyspace and transactions with revision checking. It provides weaker guarantee when accessing external resources with dependency on timing. Deadlock prevention and lock invalidation is difficult with distributed locks because a distributed lock does not protect resources within the same process, or even the same machine. To address this, distributed locks traditionally rely on a lease and heartbeat mechanism to detect disconnected lock holders and invalidate locks. When a lock holder is disconnected or paused without using revision checking (a fencing token implementation in etcd), it might access the protected resources simultaneously with the new lock holder. More details can be found [in Kleppmann's blog post](https://martin.kleppmann.com/2016/02/08/how-to-do-distributed-locking.html).
```

Comment:
"an etcd lock only safely guarantees mutual exclusion inside etcd's own keyspace and transactions with revision checking" could be a bit unclear--I first read this as guaranteeing mutual exclusion inside etcd's own keyspace, *as well as* transactions with revision checking. You might want to phrase this as "mutual exclusion for etcd transactions which check the lock revision", because operations within etcd's keyspace aren't necessarily safe when guarded by locks when they *don't* check revisions.

Reply:
> mutual exclusion for etcd transactions which check the lock revision

agree this is better.